### PR TITLE
Implementing "internal" templates

### PR DESCRIPTION
**Pull request by dantleech**
This functionality was missing from the refactoring branch.

I have changed this slightly, so that `internal` is a property of the template:

````xml
<template ...>
    <controller>...</controller>
    <view>...</view>
    <internal>true</internal>
</template>
````

Which means that configuring `internal` in the `config.yml` is now not supported - which may cause a BC break on some systems.

### CODE REVIEW

File: src/Sulu/Bundle/ContentBundle/test.php [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+class Foo
+{
+}
+
+class Bar extends Foo
+{
+}
+
+$foo = new Foo();
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/test.php]: remove

File: src/Sulu/Component/Content/Metadata/StructureMetadata.php [hunk 2]
```diff
@@ -184,4 +189,16 @@ public function getResource()
     {
         return $this->resource;
     }
+
+    /**
+     * Return if this is an internal structure
+     *
+     * TODO: Remove this: https://github.com/sulu-io/sulu/issues/1220
```
Comment (dantleech) [inline src/Sulu/Component/Content/Metadata/StructureMetadata.php]: `s/Remove/Refactor`

File: src/Sulu/Component/Content/StructureManager.php [hunk 3]
```diff
@@ -1,333 +0,0 @@
...
-use Symfony\Component\Config\ConfigCache;
-use Symfony\Component\Config\Loader\LoaderInterface;
-use Symfony\Component\Config\Resource\FileResource;
-use Symfony\Component\DependencyInjection\ContainerAware;
-
-/**
- * generates subclasses of structure to match template definitions.
- * this classes will be cached in Symfony cache.
- */
-class StructureManager extends ContainerAware implements StructureManagerInterface
```
Comment (dantleech) [inline src/Sulu/Component/Content/StructureManager.php]: This class was duplicated due to bad merge resolution.

**Comment by danrot** [comment]
Is this internal property meant to stay? If not I would not support this change...

**Comment by dantleech** [comment]
@danrot this branch needs to support "internal" one way or another. this is cleaner and the original option is very awkard to implement.

It also hides the details of "internal" from the user, as it is not part of the configuration now, which makes it easier to remove in the future.

**Comment by wachterjohannes** [comment]
:+1:

File: src/Sulu/Component/Content/Compat/Structure/StructureBridge.php [hunk 4]
```diff
@@ -214,11 +219,18 @@ public function getKey()
...
+     *
+     * {@inheritDoc}
      */
     public function getInternal()
     {
-        return false;
+        return $this->structure->isInternal();
+    }
+
+    public function setInternal($internal)
```
Comment (dantleech) [inline src/Sulu/Component/Content/Compat/Structure/StructureBridge.php]: doc

File: src/Sulu/Component/Content/Compat/Structure/StructureBridge.php [hunk 5]
```diff
@@ -214,11 +219,18 @@ public function getKey()
     }
 
     /**
-     * TODO: Implement this
+     * TODO: Remove this concept.
```
Comment (dantleech) [inline src/Sulu/Component/Content/Compat/Structure/StructureBridge.php]: Refactor + link to issue

**Comment by danrot** [comment]
@dantleech Then this option should at least optional, otherwise it will be even harder to remove than it is now (since all templates have to be changed).

**Comment by dantleech** [comment]
@danrot the existence of at least 4 elements (didn't matter which ones) was separately validated. For now I have kept the existing "not in use" string and improved the property validation. We should refactor the XML schema -- but that can be some other time.

File: src/Sulu/Component/Content/Metadata/Loader/XmlLegacyLoader.php [hunk 6]
```diff
@@ -121,8 +122,12 @@ private function loadTemplateAttributes(\DOMXPath $xpath, $type)
 
             $result = array_filter($result);
 
-            if (sizeof($result) < 4) {
-                throw new InvalidXmlException($result['key']);
+            foreach (array('key', 'view', 'controller', 'cacheLifetime') as $requiredProperty) {
+                if (!isset($result[$requiredProperty])) {
+                    throw new InvalidXmlException($type, sprintf(
+                        'Property "%s" is required in XML template file "%s"', $requiredProperty, $resource
```
Comment (dantleech) [inline src/Sulu/Component/Content/Metadata/Loader/XmlLegacyLoader.php]: Explicit property validation.

**Comment by danrot** [comment]
Ok, sounds reasonable.

**Pull request merged**